src/server/scripts/Spells/spell_mage.cpp

```diff
@@ -695,6 +701,84 @@ class spell_mage_water_elemental_freeze : public SpellScript
     }
 };
 
+// Ice Lance - 30455
+class spell_mage_ice_lance : public SpellScript
+{
+    PrepareSpellScript(spell_mage_ice_lance);
+
+    bool Validate(SpellInfo const* /*spellInfo*/) override
+    {
+        return ValidateSpellInfo({
+            SPELL_MAGE_ICE_LANCE,
+            SPELL_MAGE_ICE_LANCE_TRIGGER
+        });
+    }
+
+    void HandleOnHit(SpellEffIndex /*effIndex*/)
+    {
+        Unit* caster = GetCaster();
+        Unit* target = GetHitUnit();
+        if (!caster && !target)
+            return;
+
+        bool targetFrozen = target->HasAuraState(AURA_STATE_FROZEN);
```

Comment:
After you correct SpellClassMask as I mentioned in other comment, you will be able to stop checking if caster has SPELL_MAGE_FINGERS_OF_FROST, and only check aurastate using `target->HasAuraState(AURA_STATE_FROZEN, GetSpellInfo(), caster);`

Reply:
All done as you wrote to me.
Now when i have Finger of Frost buff, Ice Lance damage do the right job :) ...but when i dont have Finger of Frost buff and use Ice Lance on frozen target, by Frost Nova for example, the damage is not triple.

It's harder than i thought, but it almost done.

In the while i will try to learn how spells works (using also SpellWork software) in SpellEffect.db2

Reply:
That would mean that Frost Nova is not setting `AURA_STATE_FROZEN` properly